Use "fusion" style on macOS Big Sur with old Qt

hebasto opened the pull request:
The "macintosh" style is broken on macOS Big Sur:

- https://github.com/bitcoin/bitcoin/issues/20555#issuecomment-756264648
- #136

hebasto commented:
@jonasschnelli @laanwj @MarcoFalke 

As the upcoming 0.21 release is the first one after macOS Big Sur release, do you mind considering to backport this change into the 0.21 branch in order to ship the usable binaries to Big Sur users?

prusnak commented:
Why is there the version check for Qt? I see the same issue with Qt 5.12 on Big Sur.

Also checking for "macOS 11" is spurious.

Do we need the version checks at all? Why don't just apply this always on macOS?

hebasto commented:
> Why is there the version check for Qt? I see the same issue with Qt 5.12 on Big Sur.

With the latest Homebrew's Qt 5.15.2 the native macOS style looks much better.

> Also checking for "macOS 11" is spurious.

Not sure if I understand you correctly. Do you mind elaborating your comment?

> Do we need the version checks at all? Why don't just apply this always on macOS?

Many (most?) macOS users prefer the native macOS style, and the "fusion" does not look native on macOS.

jarolrod reviewed (commented):
Concept ACK
 
I applied this patch to the [rc5 source code](https://github.com/bitcoin/bitcoin/releases/tag/v0.21.0rc5) to see how it would react on macOS Big Sur 11.1 with `bitcoin-qt` built with `Qt 5.9.8`.

The name check does not work:
`
if (QSysInfo::prettyProductName().startsWith("macOS 11"))
`
Removing this check successfully allows the [Qt Fusion Style](https://doc.qt.io/qt-5/qtquickcontrols2-fusion.html) to appear. I'll investigate what's wrong with the naming.

**Branch: 9f50395735097e0687daf99098c5f72ead8abc73**
Qt Fusion Style fails to appear
![Screen Shot 2021-01-08 at 4 39 44 AM](https://user-images.githubusercontent.com/23396902/104023629-f0f9ad80-518f-11eb-92a8-6dcb25c46176.png)

**Removing The Conditional**
Fusion Style appears
![Screen Shot 2021-01-08 at 8 53 06 AM](https://user-images.githubusercontent.com/23396902/104023718-0b338b80-5190-11eb-91b5-272940d306f2.png)

hebasto commented:
Updated 9f50395735097e0687daf99098c5f72ead8abc73 -> 4e1154dfd128cbada65e9ea08ee274cdeafc4c53 ([pr177.01](https://github.com/hebasto/gui/commits/pr177.01) -> [pr177.02](https://github.com/hebasto/gui/commits/pr177.02), [diff](https://github.com/hebasto/gui/compare/pr177.01..pr177.02)):

- addressed @jarolrod's [comment](https://github.com/bitcoin-core/gui/pull/177#pullrequestreview-564281652):
> The name check does not work:
> `if (QSysInfo::prettyProductName().startsWith("macOS 11"))`

jarolrod reviewed (commented):
ACK 4e1154dfd128cbada65e9ea08ee274cdeafc4c53

Qt Fusion successfully shows. Tested with patched rc5 source code on macOS Big Sur with Qt 5.9.8

jonasschnelli commented:
This PR macOS 11 dark mode (self compiled against Qt 5.14):
<img width="1057" alt="Bildschirmfoto 2021-01-11 um 14 38 09" src="https://user-images.githubusercontent.com/178464/104189429-ee5eb880-541a-11eb-9d8a-da8824dcee59.png">
<img width="1057" alt="Bildschirmfoto 2021-01-11 um 14 38 12" src="https://user-images.githubusercontent.com/178464/104189438-f0c11280-541a-11eb-9962-4307cebe18bf.png">
<img width="1057" alt="Bildschirmfoto 2021-01-11 um 14 39 32" src="https://user-images.githubusercontent.com/178464/104189456-f880b700-541a-11eb-8b5b-4ab470344fc9.png">
<img width="638" alt="Bildschirmfoto 2021-01-11 um 14 38 18" src="https://user-images.githubusercontent.com/178464/104189559-2239de00-541b-11eb-9c8d-3bfcd677aa1b.png">
<img width="905" alt="Bildschirmfoto 2021-01-11 um 14 38 28" src="https://user-images.githubusercontent.com/178464/104189573-25cd6500-541b-11eb-953c-163d2112f967.png">

Master Dark Mode macOS 11:
<img width="1057" alt="Bildschirmfoto 2021-01-11 um 14 39 27" src="https://user-images.githubusercontent.com/178464/104189682-4a294180-541b-11eb-9ad2-be5273bd0584.png">
<img width="1057" alt="Bildschirmfoto 2021-01-11 um 14 39 32" src="https://user-images.githubusercontent.com/178464/104189695-4eedf580-541b-11eb-829c-6f275f058de3.png">
<img width="905" alt="Bildschirmfoto 2021-01-11 um 14 39 38" src="https://user-images.githubusercontent.com/178464/104189701-51e8e600-541b-11eb-909e-533b36284261.png">

I can't see any major visual differences.
Master already looks good on macOS BigSure. Is that due to Qt 5.14?

hebasto commented:
@jonasschnelli 
> Is that due to Qt 5.14?

Yes: https://github.com/bitcoin-core/gui/blob/4e1154dfd128cbada65e9ea08ee274cdeafc4c53/src/qt/bitcoin.cpp#L470-L475

jarolrod commented on src/qt/bitcoin.cpp:
```diff
@@ -466,6 +467,13 @@ int GuiMain(int argc, char* argv[])
 
+#if (QT_VERSION <= QT_VERSION_CHECK(5, 9, 8)) && defined(Q_OS_MACOS)
+    const auto os_name = QSysInfo::prettyProductName();
+    if (os_name.startsWith("macOS 11") || os_name.startsWith("macOS 10.16")) {
```
I think this is fine as is. This is meant to be a patch until the Qt version is bumped up from 5.9.8, not a permanent fix. Presumably, we will get the Qt version bumped up before there is a macOS 12.

hebasto replied:
> Perhaps it should apply to 12+ as well, though probably the issue will be long-gone by then

Did not test 10.12, but there are no visual issue on Catalina 10.15.7 (19H114). I believe the "macintosh" style is broken on macOS Big Sur only.

jarolrod commented on src/qt/bitcoin.cpp:
```diff
@@ -466,6 +467,13 @@ int GuiMain(int argc, char* argv[])
     QCoreApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
 #endif
 
+#if (QT_VERSION <= QT_VERSION_CHECK(5, 9, 8)) && defined(Q_OS_MACOS)
```
This version check is correct because #136 concerns Qt compiled on macOS with `depends`. Compiling with `depends` builds `bitcoin-qt` with `Qt 5.9.8`. See: [Qt Depends Package](https://github.com/bitcoin-core/gui/blob/master/depends/packages/qt.mk)

hebasto replied:
> Based on the linked Qt bug, this version check appears wrong?

The link to Qt bug removed from the OP. It is possible that version check could be improved, but I do not know for sure from which Qt version widgets are quite usable (I do not even say about fixing all bugs).

The recent Homebrew's Qt versions look decent (but not ideal). And this PR goal is to fix GUI look while building with depends (including gitian binary).

hebasto commented:
Btw, there is an alternative to this PR. It is reverting of https://github.com/bitcoin/bitcoin/pull/16578. Then it would be possible to pass `-style Fusion` as a command line argument.

cc @achow101

achow101 commented:
> Btw, there is an alternative to this PR. It is reverting of bitcoin/bitcoin#16578

NACK. The details will become clear soon.

MarcoFalke commented:
run appveyor (will need to be rerun before merge)

laanwj commented:
So nice to see we're finally going to support MacOS dark mode.

jonasschnelli reviewed (approved):
Tested ACK 4e1154dfd128cbada65e9ea08ee274cdeafc4c53

MarcoFalke commented:
Backported in https://github.com/bitcoin/bitcoin/pull/20901
